Logics

zipnikitin opened the pull request:
1. Сделайте назначение следующего вопроса в модели TestPassage с помощью обратного вызова -set_next_question в models/result. Не стал удалять results и создавать test_passage, так как уже использовать has_many through, добавил индексы current_question через миграцию
2. Доработайте шаблон результата прохождения теста:
Выводите сообщение и процент успеха в зависимости от количества правильных ответов - использовал таблицу стилей results.scss, обозначил вывод в двух цветах, отобразил в views/results/statistic
Добавьте навигацию в виде ссылки на список всех тестов - views/results/statistic
3. Доработайте шаблон отображения прохождения теста:
Выводите номер текущего вопроса и общее количество вопросов теста - views/results/show

zipnikitin committed:
Logics

1. Сделайте назначение следующего вопроса в модели TestPassage с помощью обратного вызова -set_next_question в models/result. Не стал удалять results и создавать test_passage, так как уже использовать has_many through, добавил индексы current_question через миграцию
2. Доработайте шаблон результата прохождения теста:
Выводите сообщение и процент успеха в зависимости от количества правильных ответов - использовал таблицу стилей results.scss, обозначил вывод в двух цветах, отобразил в views/results/statistic
Добавьте навигацию в виде ссылки на список всех тестов - views/results/statistic
3. Доработайте шаблон отображения прохождения теста:
Выводите номер текущего вопроса и общее количество вопросов теста - views/results/show

Также:
Доработал views для отображение результатов
Сделал форму для результатов
Добавил коллбэк для отображения тестов юзера
Скорректировал routes

VidgarVii commented on testGuru/app/controllers/answers_controller.rb:
```diff
@@ -0,0 +1,52 @@
+  end
+
+  def edit
+  end
```
def edit; end

VidgarVii commented on testGuru/app/controllers/answers_controller.rb:
```diff
@@ -0,0 +1,52 @@
+  def question_search
+    @question = Question.find(params[:question_id])
+  end
+
```
лишний отступ

VidgarVii commented on testGuru/app/controllers/results_controller.rb:
```diff
@@ -0,0 +1,23 @@
+  def result_search
+    @result = Result.find(params[:id])
+  end
+end
```
не тестировал. тут баг

VidgarVii commented on testGuru/app/models/result.rb:
```diff
@@ -2,5 +2,55 @@
+  end
+
+  def test_passed?
+    test_scores >= 85
```
лучше "магические цифры" хранить в константах с "говорящими именами"

VidgarVii commented on testGuru/app/models/result.rb:
```diff
@@ -2,5 +2,55 @@
+  end
+
+  def test_scores
+    (correct_questions.count * 100) / test.questions.size
```
лучше, чтоб при делении одно число было нецелым (`float`)
сейчас у тебя результат - `integer`, т.е. происходит округление до целого по умолчанию
сравни:
```
2.6.0 :001 > 3/2
 => 1
2.6.0 :002 > 3.0/2
 => 1.5
2.6.0 :003 >

VidgarVii commented on testGuru/app/models/result.rb:
```diff
@@ -2,5 +2,55 @@
+
+  def set_next_question
+    self.current_question = test.questions.order(:id).where('id > ?', current_question.id).first
+  end
```
дублирование в методах set_next_question и before_validation_set_first_question

VidgarVii commented on testGuru/app/models/result.rb:
```diff
@@ -2,5 +2,55 @@
+    correct_answers_count = correct_answers_count
+    (correct_answers_count == correct_answers.where(id: answer_ids).count) &&
+    correct_answers_count == answers_ids.count
+  end
```
почитай ДЗ. там представлен код по лучше. только надо тестировать и доработать

VidgarVii commented on testGuru/config/routes.rb:
```diff
@@ -1,7 +1,19 @@
+    end
+
+    member do
+      post :start
```
для 1 роута можно убрать блок и передать member как опцию в метод post/get

VidgarVii commented on testGuru/db/migrate/20210706221224_add_current_auestion.rb:
```diff
@@ -0,0 +1,6 @@
+class AddTestIndexLevelTittle < ActiveRecord::Migration[6.1]
+  def change
+    add_reference :results, :correct_questions, foreign_key { to_table: :questions }
```
поле для reference должно быть в единственном числе
reference это альяс belongs_to

zipnikitin committed:
Logic

Уменьшил количество строчек кода, упростив отображение методов и убрав лишние отступы
Пофиксил баг с приемом результатов и сделал float вместо integer, теперь ок
Создал константу для порога прохождения теста
Соединил методы set_next_question и before_validation_set_first_question
Взял строчку из ДЗ для проверки корректности, пофиксил
Переделал блок в роуте тестов
Создал миграцию корректно
